### ISSUE TITLE
Private error for same struct in different files

### ISSUE THREAD

**Issue by codeyash** [rust-lang/rust#43084]
field `name` of struct `school:::Name` is private when trying to split Struct into two files.

I tried this code:

```
//file **name.rs**
Struct Name{
   name: String
};

//in separate file **name_methods.rs**

impl Name{
    pub fn name() -> String{
        self.name.clone()
    }
}

```
I expected to see this happen: **No errors at all.**
Impl itself (Self ) should access its objects everywhere and doesn't depend on file structure. 

Instead, this happened:   **field `name` of struct `school:::Name` is private. Reported private error.**

## Meta

rustc 1.20.0-nightly (05b579766 2017-07-01)
binary: rustc
commit-hash: 05b5797664d6aeaa0c7d0606610f336fe0b57e97
commit-date: 2017-07-01
host: x86_64-unknown-linux-gnu
release: 1.20.0-nightly
LLVM version: 4.0

Backtrace:
not required.

**Comment by Mark-Simulacrum** [comment]
This is expected behavior. Privacy in Rust works on the basis of modules, not where `impl` is located. For this reason, the private field (and the private struct) cannot be seen from the other file, because it is another module. https://doc.rust-lang.org/nightly/reference/visibility-and-privacy.html may be a helpful reference as to possible solutions to this problem.

**Comment by codeyash** [comment]
Same mod divided into multiple files will be treated as combination of modules and can't access same mod's private fields/attr into these sub mod. Strange!

Files are merely representation of struct or mod. It shouldn't be used to identity private/pub things.

Seems odd to me.

Is there solution if anyone wants to use it like above code?

**Comment by petrochenkov** [comment]
@codeyash 
>Is there solution if anyone wants to use it like above code?

Make `name_methods` a submodule of `name`, then it will have necessary access.
```rust
mod name {
    // Finds `name_methods.rs` automatically if it's in conventional location,
    // or you can use #[path = "path/to/name_methods.rs"] explicitly, this is
    // described in the book https://doc.rust-lang.org/nightly/book/second-edition/ch07-01-mod-and-the-filesystem.html
    mod name_methods;
}
```

**Comment by Mark-Simulacrum** [comment]
Closing since this isn't a bug -- let us know if you have any more questions, though, and we'd be happy to help. 

> Same mod divided into multiple files will be treated as combination of modules and can't access same mod's private fields/attr into these sub mod. Strange!

Note that each file is implicitly a mod when declared (e.g., `mod foo;`), so "same mod" here isn't actually correct.
